README
NAME

Catalyst::Plugin::Session::Store::Redis - The great new Catalyst::Plugin::Session::Store::Redis!

SYNOPSIS

    use Catalyst qw/
        Session
        Session::Store::Redis
        Session::State::Foo
    /;
    
    MyApp->config->{session} = {
        expires => 3600,
        redis_server => '127.0.0.1:6379',
        redis_debug => 0, # or 1!
        redis_password => 'password',
        redis_reconnect => 60,
        redis_every => 5000,
    };

    # ... in an action:
    $c->session->{foo} = 'bar'; # will be saved

DESCRIPTION

Catalyst::Plugin::Session::Store::Redis is a session storage plugin for
Catalyst that uses the Redis key-value database.

NOTES

Expired Sessions

This store does B<not> automatically expire sessions.  You can call
C<delete_expired_sessions> to clear any expired sessions.  All sessions will
then be checked, one at a time.  If a session has expired then it will be
deleted.

WARNING

This module is currently untested, outside of the unit tests it ships with.
It will eventually be used with a busy site, but is currently unproven.
Patches are welcome!

AUTHOR

Cory G Watson, C<< <gphat at cpan.org> >>

COPYRIGHT & LICENSE

Copyright 2009 Cold Hard Code, LLC.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


Changes
Revision history for Catalyst-Plugin-Session-Store-Redis

0.10 March 4th, 2016
  - Allow to pass any configuration parameter to Redis
  - Force Redis at least 1.977
  - Update docs

0.06  March 8th, 2014
   - Fix disconnection Redis Server
   - Add option configuration redis_reconnect
   - Removed Catalyst log debug

0.05

0.04

0.03    December 12th, 2009
    - Fix POD description
    - Fix logging typo (said 'Getting key' when it meant 'Setting key')
    - Use Redis' EXPIRE command to build in session expiration so we can
      ignore it. As such, delete_expired_sessions is now a null-op.
    - Unify connection code that was previously duplicated in setup_session
    - Test against Redis 1.02 and 1.1.91-beta

0.02
    Keep up with changes in Catalyst-Plugin-Session

0.01
        First version, released on an unsuspecting world.



lib/Catalyst/Plugin/Session/Store/Redis.pm
package Catalyst::Plugin::Session::Store::Redis;
use warnings;
use strict;

use base qw/
    Class::Data::Inheritable
    Catalyst::Plugin::Session::Store
/;
use MRO::Compat;
use MIME::Base64 qw(encode_base64 decode_base64);
use Redis;
use Storable qw/nfreeze thaw/;
use Try::Tiny;

our $VERSION = '0.04';

__PACKAGE__->mk_classdata(qw/_session_redis_storage/);

sub get_session_data {
    my ($c, $key) = @_;

    $c->_verify_redis_connection;

    if(my ($sid) = $key =~ /^expires:(.*)/) {
        $c->log->debug("Getting expires key for $sid");
        return $c->_session_redis_storage->get($key);
    } else {
        $c->log->debug("Getting $key");
        my $data = $c->_session_redis_storage->get($key);
        if(defined($data)) {
            return thaw( decode_base64($data) )
        }
    }

    return;
}

sub store_session_data {
    my ($c, $key, $data) = @_;

    $c->_verify_redis_connection;

    if(my ($sid) = $key =~ /^expires:(.*)/) {
        $c->log->debug("Setting expires key for $sid: $data");
        $c->_session_redis_storage->set($key, $data);
    } else {
        $c->log->debug("Setting $key");
        $c->_session_redis_storage->set($key, encode_base64(nfreeze($data)));
    }

    # We use expire, not expireat because it's a 1.2 feature and as of this
    # release, 1.2 isn't done yet.
    my $exp = $c->session_expires;
    my $duration = $exp - time;
    $c->_session_redis_storage->expire($key, $duration);
    # $c->_session_redis_storage->expireat($key, $exp);

    return;
}

sub delete_session_data {
    my ($c, $key) = @_;

    $c->_verify_redis_connection;

    $c->log->debug("Deleting: $key");
    $c->_session_redis_storage->del($key);

    return;
}

sub delete_expired_sessions {
    my ($c) = @_;

    # Null op, Redis handles this for us!
}

sub setup_session {
    my ($c) = @_;

    $c->maybe::next::method(@_);

    $c->_verify_redis_connection;
}

sub _verify_redis_connection {
    my ($c) = @_;

    my $cfg = $c->_session_plugin_config;

    try {
        $c->_session_redis_storage->ping;
    } catch {
        my %redis_cfg;
        while (my ($ck, $cv) = each %$cfg) {
            next unless $ck =~ /^redis_(.+)/;
            $redis_cfg{$1} = $cv;
        }

        $redis_cfg{server} ||= '127.0.0.1:6379';

        $c->_session_redis_storage(
            Redis->new(%redis_cfg)
        );
    };
}

1;

__END__

=head1 NAME

Catalyst::Plugin::Session::Store::Redis - Redis Session store for Catalyst

=head1 SYNOPSIS

    use Catalyst qw/
        Session
        Session::Store::Redis
        Session::State::Foo
    /;
    
    MyApp->config->{Plugin::Session} = {
        expires => 3600,
        redis_server => '127.0.0.1:6379',
        redis_debug => 0, # or 1!
        redis_password => 'password',
        redis_reconnect => 60,
        redis_every => 5000,
    };

    # ... in an action:
    $c->session->{foo} = 'bar'; # will be saved

=head1 DESCRIPTION

C<Catalyst::Plugin::Session::Store::Redis> is a session storage plugin for
Catalyst that uses the Redis (L<http://code.google.com/p/redis/>) key-value
database.

=head1 NOTES

=over 4

=item B<Expired Sessions>

This store does B<not> automatically expire sessions.  You can call
C<delete_expired_sessions> to clear any expired sessions.  All sessions will
then be checked, one at a time.  If a session has expired then it will be
deleted.

=back

=head1 WARNING

This module is currently untested, outside of the unit tests it ships with.
It will eventually be used with a busy site, but is currently unproven.
Patches are welcome!

=head1 AUTHOR

Cory G Watson, C<< <gphat at cpan.org> >>

=head1 COPYRIGHT & LICENSE

Copyright 2009 Cold Hard Code, LLC.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut
